### PR TITLE
Add `BloomModel` hydra support

### PR DESCRIPTION
**Pull request by jon-tow**
This PR adds hydra-based PPO model branching support for `BloomModel`s.

`wandb` reports:
* [PPO Sentiment](https://wandb.ai/jon-tow/bloom-trlx/reports/trlx-BLOOM-Hydra-Support-Sanity-Benchmark--VmlldzozMTExNzQx?accessToken=di165xkctf625t8szihzw87qpbnb71qjxit92dxjhkc4kbiragjnpoaxygt62k6a)

Note: The `<Arch>ModelBranch` implementations should be refactored in the future to share a common interface if we plan on adding more in the near future.

### CODE REVIEW

**Review by LouisCastricato — APPROVED** [review]

**Comment by LouisCastricato** [comment]
@Dahoas can you give this a quick glance

**Comment by Dahoas** [comment]
Bloom rewards seem a bit low relative to other models :(. But Not much we can really do I guess

File: trlx/model/nn/ppo_models.py [hunk 1]
```diff
@@ -798,6 +799,193 @@ def forward(
...
+        batch_size, seq_length = hidden_states.shape[:2]
+
+        if past_key_values is None:
+            past_key_values = tuple([None] * len(self.transformer_blocks))
+
+        # Prepare head mask if needed
+        # 1.0 in head_mask indicate we keep the head
+        # attention_probs has shape batch_size x num_heads x N x N
+        # head_mask has shape n_layer x batch x num_heads x N x N
+        head_mask = self.get_head_mask(head_mask, hf_get_num_hidden_layers(self.config))
```
Comment (Dahoas) [inline trlx/model/nn/ppo_models.py:879]: What is the head mask?

File: trlx/model/nn/ppo_models.py [hunk 2]
```diff
@@ -798,6 +799,193 @@ def forward(
...
+            past_key_values_length = past_key_values[0][0].shape[2]
+            seq_length_with_past = seq_length_with_past + past_key_values_length
+        if attention_mask is None:
+            attention_mask = torch.ones(
+                (batch_size, seq_length_with_past), device=hidden_states.device
+            )
+        else:
+            attention_mask = attention_mask.to(hidden_states.device)
+
+        alibi = modeling_bloom.build_alibi_tensor(
```
Comment (Dahoas) [inline trlx/model/nn/ppo_models.py:898]: I didn't know bloom used alibi

**Comment by LouisCastricato** [comment]
To he expected

**Review by Dahoas — APPROVED** [review]

**Comment by Dahoas** [comment]
Looks good to me! I just left some questions so I can better understand how things are functioning.

File: trlx/model/nn/ppo_models.py (hunk 1 above)
Comment (jon-tow) [reply trlx/model/nn/ppo_models.py:879]: The head mask is a binary mask that can be used to drop the self-attention weights (`softmax(qk)`) from specified heads before computing the full attention output. For example, see [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/gptj/modeling_gptj.py#L190-192). [`get_head_mask`](https://github.com/huggingface/transformers/blob/31d452c68b34c2567b62924ee0df40a83cbc52d5/src/transformers/modeling_utils.py#L799) just expands the dimensions to line-up with proper shape.
